File: backend/monitor/openapi/observations.py
"""OpenAPI schemas for observations, FAST details, and trajectories."""

from __future__ import annotations

from .common import _nullable


def observation_schemas(nullable_number: dict, nullable_string: dict) -> dict:
    return {
        "ObservationList": {
            "type": "object",
            "required": [
                "account",
                "items",
                "corrections_available",
                "pagination",
                "summary",
            ],
            "properties": {
                "account": {
                    "oneOf": [
                        {
                            "$ref": "#/components/schemas/MonitoredAccountSummary"
                        },
                        {"type": "null"},
                    ]
                },
                "items": {
                    "type": "array",
                    "items": {"$ref": "#/components/schemas/Observation"},
                },
                "corrections_available": {"type": "boolean"},
                "pagination": {
                    "$ref": "#/components/schemas/Pagination"
                },
                "summary": {
                    "type": "object",
                    "required": [
                        "total",
                        "valid_count",
                        "passive_count",
                        "excluded_count",
                    ],
                    "properties": {
                        "total": {"type": "integer"},
                        "valid_count": {"type": "integer"},
                        "passive_count": {"type": "integer"},
                        "excluded_count": {"type": "integer"},
                    },
                },
            },
        },
        "Observation": {
            "type": "object",
            "required": [
                "id",
                "observed_at",
                "source",
                "account_id",
                "attribution_started_at",
                "upstream_resets_at",
                "upstream_used_percent",
                "interval_used_percent",
                "raw_selected_total_cost",
                "selected_total_cost",
                "cost_window_started_at",
                "cost_window_ended_at",
                "interval_cost_started_at",
                "interval_cost",
                "interval_cost_source",
                "normalized_total_cost",
                "delta_percent",
                "delta_cost",
                "sample_usd_per_percent",
                "effective_usd_per_percent",
                "estimated_used_percent",
                "capacity_lower_usd",
                "capacity_upper_usd",
                "model_diagnostics",
                "fast_correction_usd",
                "fast_correction_calculated",
                "valid_sample",
                "sample_note",
                "rate_method",
                "query_mode",
                "snapshot_sampled_at",
                "excluded",
                "excluded_at",
                "exclusion_reason",
                "exclusion_source",
                "is_manual_start",
                "manual_start_reason",
                "manual_start_set_at",
                "manual_start_end_id",
                "manual_start_end_observed_at",
                "participants",
            ],
            "properties": {
                "id": {"type": "integer"},
                "observed_at": {"type": "string", "format": "date-time"},
                "source": {
                    "type": "string",
                    "enum": [
                        "scheduled",
                        "manual",
                        "exhausted",
                        "reset",
                    ],
                },
                "account_id": {
                    "type": "integer",
                    "description": "Sub2API 上游账号 ID。",
                },
                "attribution_started_at": nullable_string,
                "upstream_resets_at": {
                    "type": "string",
                    "format": "date-time",
                },
                "upstream_used_percent": {"type": "number"},
                "interval_used_percent": {"type": "number"},
                "raw_selected_total_cost": {"type": "number"},
                "selected_total_cost": {"type": "number"},
                "cost_window_started_at": nullable_string,
                "cost_window_ended_at": nullable_string,
                "interval_cost_started_at": nullable_string,
                "interval_cost": nullable_number,
                "interval_cost_source": {"type": "string"},
                "normalized_total_cost": {"type": "number"},
                "delta_percent": nullable_number,
                "delta_cost": nullable_number,
                "sample_usd_per_percent": nullable_number,
                "effective_usd_per_percent": {"type": "number"},
                "estimated_used_percent": {"type": "number"},
                "capacity_lower_usd": nullable_number,
                "capacity_upper_usd": nullable_number,
                "model_diagnostics": {"type": "object"},
                "fast_correction_usd": nullable_number,
                "fast_correction_calculated": {"type": "boolean"},
                "valid_sample": {"type": "boolean"},
                "sample_note": {"type": "string"},
                "rate_method": {"type": "string"},
                "query_mode": {
                    "type": "string",
                    "enum": ["passive", "direct"],
                },
                "snapshot_sampled_at": nullable_string,
                "excluded": {"type": "boolean"},
                "excluded_at": nullable_string,
                "exclusion_reason": {"type": "string"},
                "exclusion_source": {
                    "type": "string",
                    "enum": ["", "manual", "automatic"],
                },
                "is_manual_start": {"type": "boolean"},
                "manual_start_reason": {"type": "string"},
                "manual_start_set_at": nullable_string,
                "manual_start_end_id": _nullable("integer"),
                "manual_start_end_observed_at": nullable_string,
                "participants": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/ParticipantSnapshot"
                    },
                },
            },
        },
        "FastCorrectionDetail": {
            "type": "object",
            "required": [
                "observation_id",
                "started_at",
                "ended_at",
                "calculated",
                "cost_basis",
                "cost_basis_label",
                "request_count",
                "fast_request_count",
                "non_fast_request_count",
                "fast_billed_cost_usd",
                "correction_usd",
                "corrected_fast_cost_usd",
                "collection_error",
                "users",
            ],
            "properties": {
                "observation_id": {"type": "integer"},
                "started_at": nullable_string,
                "ended_at": {"type": "string", "format": "date-time"},
                "calculated": {"type": "boolean"},
                "cost_basis": {
                    "type": "string",
                    "enum": ["actual", "standard"],
                },
                "cost_basis_label": {"type": "string"},
                "request_count": _nullable("integer"),
                "fast_request_count": {"type": "integer"},
                "non_fast_request_count": _nullable("integer"),
                "fast_billed_cost_usd": {"type": "number"},
                "correction_usd": {"type": "number"},
                "corrected_fast_cost_usd": {"type": "number"},
                "collection_error": {"type": "string"},
                "users": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/FastCorrectionUser"
                    },
                },
            },
        },
        "FastCorrectionUser": {
            "type": "object",
            "required": [
                "sub2api_user_id",
                "username",
                "email",
                "display_name",
                "request_count",
                "fast_request_count",
                "non_fast_request_count",
                "fast_billed_cost_usd",
                "correction_usd",
                "corrected_fast_cost_usd",
            ],
            "properties": {
                "sub2api_user_id": {"type": "integer"},
                "username": {"type": "string"},
                "email": {"type": "string"},
                "display_name": {"type": "string"},
                "request_count": _nullable("integer"),
                "fast_request_count": {"type": "integer"},
                "non_fast_request_count": _nullable("integer"),
                "fast_billed_cost_usd": {"type": "number"},
                "correction_usd": {"type": "number"},
                "corrected_fast_cost_usd": {"type": "number"},
            },
        },
        "ParticleTrajectory": {
            "type": "object",
            "required": ["available", "message"],
            "properties": {
                "account": {
                    "$ref": "#/components/schemas/MonitoredAccountSummary"
                },
                "available": {"type": "boolean"},
                "message": {"type": "string"},
                "algorithm": {"type": "string"},
                "seed": {"type": "integer"},
                "particle_count": {"type": "integer"},
                "representative_particle_count": {"type": "integer"},
                "credible_mass_percent": {"type": "number"},
                "selected_period_id": {"type": "integer"},
                "periods": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/TrajectoryPeriod"
                    },
                },
                "segment": {"type": "object"},
                "cycle_usage": {
                    "$ref": "#/components/schemas/TrajectoryCycleUsage"
                },
                "latest": {"type": "object"},
                "points": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/TrajectoryPoint"
                    },
                },
                "promotions": {
                    "type": "array",
                    "items": {
                        "$ref": "#/components/schemas/RangePromotion"
                    },
                },
            },
        },
        "TrajectoryCycleUsage": {
            "type": "object",
            "required": [
                "observed_at",
                "account_total_usd",
                "estimated_used_percent",
                "displayed_used_percent",
                "participants",
            ],
            "properties": {
                "observed_at": {"type": "string", "format": "date-time"},
                "account_total_usd": {"type": "number"},
                "estimated_used_percent": {"type": "number"},
                "displayed_used_percent": {"type": "number"},
                "participants": {
                    "type": "array",
                    "items": {
                        "$ref": (
                            "#/components/schemas/"
                            "TrajectoryParticipantUsage"
                        )
                    },
                },
            },
        },
        "TrajectoryParticipantUsage": {
            "type": "object",
            "required": [
                "participant_id",
                "participant_name",
                "is_owner",
                "used_usd",
            ],
            "properties": {
                "participant_id": {"type": "integer"},
                "participant_name": {"type": "string"},
                "is_owner": {"type": "boolean"},
                "used_usd": {"type": "number"},
            },
        },
        "TrajectoryPeriod": {
            "type": "object",
            "required": [
                "id",
                "sequence",
                "started_at",
                "first_observed_at",
                "last_observed_at",
                "resets_at",
                "ended_at",
                "observation_count",
                "is_current",
            ],
            "properties": {
                "id": {"type": "integer"},
                "sequence": {"type": "integer"},
                "started_at": {"type": "string", "format": "date-time"},
                "first_observed_at": {
                    "type": "string",
                    "format": "date-time",
                },
                "last_observed_at": {
                    "type": "string",
                    "format": "date-time",
                },
                "resets_at": {"type": "string", "format": "date-time"},
                "ended_at": {"type": "string", "format": "date-time"},
                "observation_count": {"type": "integer"},
                "is_current": {"type": "boolean"},
            },
        },
        "TrajectoryPoint": {
            "type": "object",
            "required": [
                "observation_id",
                "observed_at",
                "source",
                "displayed_percent",
                "estimated_percent",
                "estimated_percent_lower",
                "estimated_percent_upper",
                "capacity_usd",
                "capacity_lower_usd",
                "capacity_upper_usd",
                "range_min_usd",
                "range_max_usd",
                "range_stage",
                "range_direction",
                "ess_fraction",
                "resampled",
                "boundary_mass",
                "particles_usd",
            ],
            "properties": {
                "observation_id": {"type": "integer"},
                "observed_at": {"type": "string", "format": "date-time"},
                "source": {"type": "string"},
                "displayed_percent": {"type": "number"},
                "estimated_percent": {"type": "number"},
                "estimated_percent_lower": {"type": "number"},
                "estimated_percent_upper": {"type": "number"},
                "capacity_usd": {"type": "number"},
                "capacity_lower_usd": {"type": "number"},
                "capacity_upper_usd": {"type": "number"},
                "range_min_usd": {"type": "number"},
                "range_max_usd": {"type": "number"},
                "range_stage": {"type": "integer"},
                "range_inherited": {"type": "boolean"},
                "range_direction": {
                    "type": ["string", "null"],
                    "enum": ["upper", "lower", None],
                },
                "ess_fraction": {"type": "number"},
                "resampled": {"type": "boolean"},
                "boundary_mass": {
                    "type": "object",
                    "required": ["lower", "upper"],
                    "properties": {
                        "lower": {"type": "number"},
                        "upper": {"type": "number"},
                    },
                },
                "particles_usd": {
                    "type": "array",
                    "items": {"type": "number"},
                },
            },
        },
        "RangePromotion": {
            "type": "object",
            "required": [
                "stage",
                "direction",
                "occurred_at",
                "from_range_usd",
                "to_range_usd",
                "boundary_mass",
                "display_residual_pp",
            ],
            "properties": {
                "stage": {"type": "integer"},
                "direction": {
                    "type": "string",
                    "enum": ["upper", "lower"],
                },
                "occurred_at": {"type": "string", "format": "date-time"},
                "from_range_usd": {
                    "type": "array",
                    "prefixItems": [
                        {"type": "number"},
                        {"type": "number"},
                    ],
                    "minItems": 2,
                    "maxItems": 2,
                },
                "to_range_usd": {
                    "type": "array",
                    "prefixItems": [
                        {"type": "number"},
                        {"type": "number"},
                    ],
                    "minItems": 2,
                    "maxItems": 2,
                },
                "boundary_mass": {"type": "number"},
                "display_residual_pp": {"type": "number"},
            },
        },
    }

File: frontend/src/types/observations.ts
import type { MonitoredAccount } from "./accounts";
import type {
  CorrectionBreakdown,
  CorrectionSource,
  PaginatedData,
} from "./common";
import type { ModelDiagnostics } from "./dashboard";
import type { Snapshot } from "./participants";

export interface Observation extends CorrectionBreakdown {
  id: number;
  observed_at: string;
  source: string;
  provider: "sub2api" | "cpa" | "gpt_load";
  account_id: number;
  correction_source: CorrectionSource;
  pricing_epoch: string;
  attribution_started_at: string | null;
  upstream_resets_at: string;
  upstream_used_percent: number;
  interval_used_percent: number;
  raw_selected_total_cost: number;
  selected_total_cost: number;
  cost_window_started_at: string | null;
  cost_window_ended_at: string | null;
  interval_cost_started_at: string | null;
  interval_cost: number | null;
  interval_cost_source: string;
  normalized_total_cost: number;
  delta_percent: number | null;
  delta_cost: number | null;
  sample_usd_per_percent: number | null;
  effective_usd_per_percent: number;
  estimated_used_percent: number;
  capacity_lower_usd: number | null;
  capacity_upper_usd: number | null;
  model_diagnostics: ModelDiagnostics | Record<string, never>;
  fast_correction_usd: number | null;
  fast_correction_calculated: boolean;
  valid_sample: boolean;
  sample_note: string;
  rate_method: string;
  query_mode: string;
  snapshot_sampled_at: string | null;
  participants: Snapshot[];
  excluded: boolean;
  excluded_at: string | null;
  exclusion_reason: string;
  exclusion_source: "" | "manual" | "automatic";
  is_manual_start: boolean;
  manual_start_reason: string;
  manual_start_set_at: string | null;
  manual_start_end_id: number | null;
  manual_start_end_observed_at: string | null;
}
export interface FastCorrectionUserDetail extends CorrectionBreakdown {
  raw_cost_usd?: number | null;
  corrected_cost_usd?: number | null;
  sub2api_user_id: number;
  username: string;
  email: string;
  display_name: string;
  request_count: number | null;
  fast_request_count: number;
  non_fast_request_count: number | null;
  fast_billed_cost_usd: number;
  correction_usd: number;
  corrected_fast_cost_usd: number;
}
export interface FastCorrectionCalculateResult extends CorrectionBreakdown {
  observation_id: number;
  fast_correction_usd: number;
  fast_correction_calculated: boolean;
  correction_calculated?: boolean;
}
export type FastCorrectionDetail =
  | HistoricalCorrectionDetail
  | {
      observation_id: number;
      correction_source: "upstream" | "none";
      pricing_epoch: string;
      message: string;
    };
export interface HistoricalCorrectionDetail extends CorrectionBreakdown {
  correction_source: "local";
  pricing_epoch: string;
  raw_cost_usd?: number | null;
  corrected_cost_usd?: number | null;
  rules_digest?: string;
  rules?: Record<string, unknown>;
  model_details?: CorrectionModelDetail[];
  observation_id: number;
  started_at: string | null;
  ended_at: string;
  calculated: boolean;
  cost_basis: "actual" | "standard";
  cost_basis_label: string;
  request_count: number | null;
  fast_request_count: number;
  non_fast_request_count: number | null;
  fast_billed_cost_usd: number;
  correction_usd: number;
  corrected_fast_cost_usd: number;
  collection_error: string;
  users: FastCorrectionUserDetail[];
}
export interface ObservationRebuildResult {
  rebuilt_observations: number;
  automatic_exclusions: number;
  inferred_intervals: number;
  latest_observation_id: number | null;
  replay_started_at: string | null;
}
export interface MonitorSchedule {
  monitoring_enabled: boolean;
  interval_seconds: number;
  next_local_check_at: string | null;
  run_in_progress: boolean;
  accounts: Array<{
    id: number;
    provider: "sub2api" | "cpa" | "gpt_load";
    source_account_id: string;
    external_account_id: number | null;
    name: string;
    enabled: boolean;
    next_local_check_at: string | null;
    run_in_progress: boolean;
  }>;
  server_time: string;
}
export interface ObservationListData extends PaginatedData<Observation> {
  account: Pick<
    MonitoredAccount,
    "id" | "provider" | "source_account_id" | "external_account_id" | "name"
  > | null;
  corrections_available: boolean;
  summary: {
    total: number;
    valid_count: number;
    passive_count: number;
    excluded_count: number;
  };
}

export interface CorrectionModelDetail extends CorrectionBreakdown {
  model: string;
  service_tier: string;
  request_count: number;
  raw_cost_usd: number;
  corrected_cost_usd: number;
  fast_factor: string;
  long_context_factor: string;
  model_factor: string;
  long_context_evidence: string;
}
